[C10D]  Support group_dst/group_src in c10d send/recv object_list

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #140847
* #140843

Also add mypy annotations

Partially addresses RFC 0042 (https://github.com/pytorch/rfcs/pull/71)
See more details/motivation in https://github.com/pytorch/pytorch/pull/140460


cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

wconstab committed:
[C10D]  Support group_dst/group_src in c10d send/recv object_list

[ghstack-poisoned]

wconstab committed:
Update on "[C10D]  Support group_dst/group_src in c10d send/recv object_list"




[ghstack-poisoned]

wconstab committed:
Update on "[C10D]  Support group_dst/group_src in c10d send/recv object_list"




[ghstack-poisoned]

wconstab committed:
Update on "[C10D]  Support group_dst/group_src in c10d send/recv object_list"




[ghstack-poisoned]

wconstab committed:
Update on "[C10D]  Support group_dst/group_src in c10d send/recv object_list"




[ghstack-poisoned]

wconstab committed:
Update on "[C10D]  Support group_dst/group_src in c10d send/recv object_list"




[ghstack-poisoned]

wconstab committed:
Update on "[C10D]  Support group_dst/group_src in c10d send/recv object_list"




[ghstack-poisoned]

wconstab committed:
Update on "[C10D]  Support group_dst/group_src in c10d send/recv object_list"


Also add mypy annotations

Partially addresses RFC 0042 (https://github.com/pytorch/rfcs/pull/71)
See more details/motivation in https://github.com/pytorch/pytorch/pull/140460


[ghstack-poisoned]

H-Huang commented on torch/distributed/distributed_c10d.py:
```diff
@@ -3105,7 +3111,8 @@ def send_object_list(object_list, dst, group=None, device=None):
             serialized and converted to tensors which are moved to the
             ``device`` before sending. Default is ``None``.
-
+        group_dst (int, optional): Destination rank on ``group``.
```
if you specify `group_dst` is there a requirement to also specify `group`?

wconstab replied:
no. Technically, group defaults to 'default_group' and that's documented.  So I allowed using group_dst along with default group.  (This means in that case group_dst would be == global_dst).

H-Huang commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -3940,12 +3943,22 @@ def test_send_recv_object_list_subgroup(self, set_device: SetDeviceMethod):
             x = [{}]
-            c10d.recv_object_list(x, src=self.rank + 1, group=subgroup, device=device)
+            if group_rank:
+                c10d.recv_object_list(x, group_src=1, group=subgroup, device=device)
```
instead of `group_src=1` is there a way to do something like `subgroup.local_rank() + 1`?

wconstab replied:
yea, it would be equivalent. I just left it this way bc the test already hardcodes 4 ranks and 2 subgroups, so this is pretty clear IMO

H-Huang reviewed (approved):
looks good

wconstab committed:
Update on "[C10D]  Support group_dst/group_src in c10d send/recv object_list"


Also add mypy annotations

Partially addresses RFC 0042 (https://github.com/pytorch/rfcs/pull/71)
See more details/motivation in https://github.com/pytorch/pytorch/pull/140460


[ghstack-poisoned]

wconstab commented:
@pytorchbot merge
